### PR TITLE
[MUGEN] Add MultimodalGPT Module

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #257
* #264

Summary:
- Defines the model architecture for the full multimodal GPT as the basis for the builder
- Defines the API for integration with generation utility
- Added `latent_shape` to reshape the token ids for decoding back to the real data
- Pulled token embedding layers out of `MultimodalTransformerDecoder` and put in `MultimodalGPT`.

Test Plan:

```
$ python -m pytest --cov=torchmultimodal/models/ test/models/test_gpt.py -vv
================================================= test session starts ==================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/t2v/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 20 items

test/models/test_gpt.py::TestMultimodalGPT::test_tokenizers_missing_methods PASSED       [  3%]
test/models/test_gpt.py::TestMultimodalGPT::test_encode_invalid_modality PASSED          [  7%]
test/models/test_gpt.py::TestMultimodalGPT::test_decode_tokens_wrong_shape PASSED        [ 11%]
test/models/test_gpt.py::TestMultimodalGPT::test_decode_tokens_reshape PASSED            [ 15%]
test/models/test_gpt.py::TestMultimodalGPT::test_lookup_invalid_modality PASSED          [ 19%]
test/models/test_gpt.py::TestMultimodalGPT::test_lookup_in_modality PASSED               [ 23%]
test/models/test_gpt.py::TestMultimodalGPT::test_lookup_out_modality PASSED              [ 26%]
test/models/test_gpt.py::TestMultimodalGPT::test_fwd_bad_input PASSED                    [ 30%]
test/models/test_gpt.py::TestMultimodalGPT::test_fwd_for_generation PASSED               [ 34%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward PASSED                          [ 38%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_logits_mask PASSED              [ 42%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_bad_input PASSED         [ 46%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_in_modality PASSED [ 50%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_out_modality PASSED [ 53%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_two_modality PASSED [ 57%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_eval_right_shift_on PASSED [ 61%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_forward_eval_right_shift_off PASSED [ 65%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_bad_pos_ids PASSED       [ 69%]
test/models/test_gpt.py::TestMultimodalTransformerDecoder::test_optional_pos_ids PASSED  [ 73%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward PASSED                     [ 76%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward_additional_output PASSED   [ 80%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward PASSED                [ 84%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_masked PASSED         [ 88%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_additional_output PASSED [ 92%]
test/models/test_gpt.py::test_sigmoid_linear_unit PASSED                                 [ 96%]
test/models/test_gpt.py::test_right_shift PASSED                                         [100%]

---------- coverage: platform darwin, python 3.8.13-final-0 ----------
Name                                            Stmts   Miss  Cover
-------------------------------------------------------------------
torchmultimodal/models/gpt.py                     181      4    98%


==== 26 passed in 1.80s =======
```

Differential Revision: [D38642048](https://our.internmc.facebook.com/intern/diff/D38642048)

### CODE REVIEW

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:105]: I don't see `num_tokens` and `d_model` used elsewhere except for initializing the head, do we need to keep them as attributes?

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:122]: I know we've been consistent with the name `head_mask`, but it may be confusing here since you have an actual head. maybe consider renaming to `attn_head_mask` or similar? your call

File: test/models/test_gpt.py
Comment (RdoubleA) [inline test/models/test_gpt.py:312]: no assertion here?

File: test/models/test_gpt.py
Comment (RdoubleA) [inline test/models/test_gpt.py:198]: could also add test for the ValueErrors in encode and decode

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [inline torchmultimodal/models/gpt.py:182]: layernorm is usually not with the head (ex: VisionTransformer from torchvision [puts it with encoder](https://github.com/pytorch/vision/blob/cc0a83856c979f834132a5faf71a2eece0d32dc3/torchvision/models/vision_transformer.py#L157), and FLAVA has it separated, [image](https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/models/flava/flava_image_encoder.py#L269) and [text](https://github.com/facebookresearch/multimodal/blob/f9327dec04a5e8c824fba36e4ed52528e44541f6/torchmultimodal/models/flava/flava_text_encoder.py#L218) encoder have a separate `Fp32LayerNorm` from the `Pooler` MLP head. I am not entirely sure if this is a convention, but maybe you could consider moving it out of `head` into `forward` or `fwd`.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:59]: that's right. We are doing self attention here so query and key sequences are the same. Will edit.

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:129]: >During training:
- encode converts in- and out-modality data to quantized embeddings
- forward takes the quantized embeddings (concatenated) for training

I will move `encode` and add concatenation step to `forward`.

For generation, `forward` is not used as we have to use those components in separation except for `MultimodalTransformerDecoder` (that operates at token level).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:129]: >I will move encode and add concatenation step to forward.

Ended up not doing this change: `encode` returns both the quantized embeddings and tokens. The latter are in fact "labels" (or ground truth) during training of the GPT decoder. Since we don't want to compute loss inside `forward`, I will keep `encode` and `forward` separate and line them up in the lightning module.

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:59]: Ended up not doing this. 

Because self attention mask requires `1`s only in the lower triangle of the matrix ``(query_length, key_length)`` (due to causal) so it cannot be broadcasted from either query or key dim (if only one is given).

- The two dimensional case is only for padding masks
- The three dimensional case is of self attention masks

https://github.com/facebookresearch/multimodal/blob/main/torchmultimodal/utils/attention.py#L30-L38

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:48]: I will just add what the default module is and the user can check the corresponding API to avoid repetition.

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:105]: They are used by `LayerNorm` and `to_logit`: 

```
self.norm = nn.LayerNorm(d_model)
self.to_logit = nn.Linear(d_model, num_tokens, bias=False)
self.to_logit.weight.data.copy_(torch.zeros(self.num_tokens, self.d_model))
```

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:122]: Not sure if I understand the proposed confusion here. 
`head_mask` is just padding mask for each attention head. 
For now we can just be consistent with the rest of the codebase: https://github.com/facebookresearch/multimodal/search?q=head_mask

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:182]: LayerNorm and logit layer are attached to the transformer decoder trunk during [training](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/gpt.py#L109-L114). However, they need to be [detached](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/gpt/gpt.py#L219-L226) from the trunk] when learning the key/value inner representation for the in-modality and then [attached](https://github.com/mugen-org/MUGEN_baseline/blob/783b5666186cae9e0517209925f0021db4026c19/lib/models/gpt/gpt.py#L227-L229) again when generating the next out-modality token. 

Grouping them together behind a method makes it flexible to attach/detach as needed.

The examples you cited are TransformerEncoder which may also be applicable for TransformerDecoder if the latter does not need to support AR decoding.

File: test/models/test_gpt.py
Comment (langong347) [reply test/models/test_gpt.py:312]: Added. Also added a separate test for the `logits_mask`.

File: test/models/test_gpt.py
Comment (langong347) [reply test/models/test_gpt.py:198]: Added.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [reply torchmultimodal/models/gpt.py:104]: right, I was just thinking you could use `d_model` and `num_tokens` directly and retire `self.d_model` and `self.num_tokens`

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [reply torchmultimodal/models/gpt.py:121]: This `head_mask` is in reference to the attention head but you also have the logits head, so I was just wondering if we should clarify that this is for the attention head and not the logits head to avoid confusion

File: torchmultimodal/models/gpt.py
Comment (RdoubleA) [reply torchmultimodal/models/gpt.py:181]: ok I see, makes sense

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:121]: I see. you meant `head` could be confused with `head` for the logit prediction.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:435]: If that's the case, should we reflect it in the type info of the init method?

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:167]: Question about how we are splitting things up here. It feels unintuitive to have a separate `fwd` method that is doing most, but not all, of the usual forward. I was gonna suggest calling it decoder or something, but that is already an overloaded term here. Wondering if it makes sense to just add a util method that will return both in and out modality (where one can be None), then call `self.mm_decoder` in the main forward method. Then you don't have to pass all these args, and the core logic is clear in forward.

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:229]: Maybe avoid the inplace version (unless you have a good reason for it). [Ref](https://lernapparat.de/pytorch-inplace)

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:128]: nit: could just call this logit_projection or something

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:234]: Another option for this (and decode method): just pass the desired tokenizer directly, then you don't need to pass the modality string.

File: torchmultimodal/models/gpt.py
Comment (ebsmothers) [inline torchmultimodal/models/gpt.py:86]: One question here: do most tokenizers implement `lookup` method? Personally I haven't seen it before.

File: test/models/test_gpt.py
Comment (ebsmothers) [inline test/models/test_gpt.py:192]: Not sure if this is the right place for it or not, but it might be worthwhile to test that things work with a nontrivial tokenizer somewhere as well.

File: test/models/test_gpt.py
Comment (ebsmothers) [inline test/models/test_gpt.py:389]: Any reason for explicitly calling `.forward` here?

File: test/models/test_gpt.py
Comment (ebsmothers) [inline test/models/test_gpt.py:435]: same question here

**Review by ebsmothers — APPROVED** [review]
Left a bunch of minor comments, but no major concerns from me. Accepting to unblock you

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:128]: I will keep it this way for now as it will make my life easier when re-mapping MUGEN's checkpoints. :)

File: torchmultimodal/models/gpt.py
Comment (langong347) [reply torchmultimodal/models/gpt.py:86]: Let me add a reference to VQVAE API since it has the lookup method.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38642048).

**Pull request closed without merging**
